### PR TITLE
Mint NOIDs when releasing metadata

### PR DESCRIPTION
**Pull request by rschenk**
@cam156 We took your advice and minted NOIDs to be used as a public identifier for Submissions once they get released. We're storing it in the `submissions.public_id` column of the db. We based the implementation off what you folks did in [psu-stewardship/active_fedora-noid](https://github.com/psu-stewardship/active_fedora-noid).

One thing we did differently was allow the minter method to accept a block used to ensure the NOID's uniqueness, so the minter isn't coupled to the model that it's minting NOIDs for. I wanted to specifically get your feedback on whether or not we should use Mutexes for multi-thread/-process safety. Given the way this particular app releases Submissions, chances are there won't be two people minting NOIDs concurrently. That being said, if you think it's worth doing, we'll definitely add it—just wanted to ask first.

### CODE REVIEW

File: app/models/noid_minter.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  attr_accessor :template
+
+  def initialize(template: default_template)
+    @template = template
+  end
+
+  # Accepts a optional block, which is passed a candidate for the new noid. The
+  # block must return true if the candidate noid already exists, false otherwise.
+  def mint(&noid_exists)
+    loop do
```
Comment (rschenk) [inline app/models/noid_minter.rb]: @cam156 Do you think it'd be worth surrounding this with a Mutex for concurrency safety?

File: app/models/submission.rb [hunk 2]
```diff
@@ -210,10 +210,11 @@ def self.release_for_publication submission_ids, date_to_release
         new_publication_release_date = s.publication_release_date date_to_release
         new_metadata_release_date = s.released_metadata_at.nil? ? date_to_release : s.released_metadata_at
         new_access_level = s.publication_release_access_level
+        new_public_id = s.public_id.presence || NoidMinter.new.mint{ |candidate| self.find_by(public_id: candidate).present? }
```
Comment (rschenk) [inline app/models/submission.rb]: The model that's creating a new NOID is responsible for checking its uniqueness via an optional block

**Comment by cam156** [comment]
@rschenk Given that the system is in control of when a noid is minted it is less likely to have an issue.  We generally get conflicts when users are uploading many spall files, which get a noid minted for them in many separate ruby threads.
I think for the first generation we should keep it simple.  Are there database checks on the column to be certain the value is unique?

**Comment by rschenk** [comment]
@cam156 Regarding keeping it simple, that was my thoughts as well so I'm glad you agree. And yes, there are database checks on that column to ensure the minted noid is unique in a non-threadsafe/-concurrent way, so I think we're OK for a first pass. :+1: Thanks!

**Pull request merged**